kamu-adapter-http/tests/harness/server_side_harness.rs

```diff
@@ -136,25 +137,27 @@ struct TestAPIServer {
 
 impl TestAPIServer {
     pub fn new(catalog: Catalog, address: Option<IpAddr>, port: Option<u16>) -> Self {
-        let local_repo: Arc<dyn DatasetRepository> = catalog.get_one().unwrap();
-        let local_dataset_resolver: Arc<dyn kamu_adapter_http::DatasetResolver> = Arc::new(
-            kamu_adapter_http::DatasetResolverLocalRepository::new(local_repo),
-        );
+        use axum::extract::Path;
+
+        #[derive(serde::Deserialize)]
+        struct DatasetByName {
+            dataset_name: DatasetName,
+        }
 
         let app = axum::Router::new()
             .nest(
-                format!("/:{}", kamu_adapter_http::PARAMETER_DATASET_NAME).as_str(),
-                kamu_adapter_http::create_dataset_transfer_protocol_routes()
-                    .layer(axum::middleware::from_fn(
-                        kamu_adapter_http::resolve_dataset_by_name,
+                "/:dataset_name",
+                kamu_adapter_http::smart_transfer_protocol_routes()
+                    .layer(kamu_adapter_http::DatasetResolverLayer::new(
+                        |Path(p): Path<DatasetByName>| p.dataset_name.as_local_ref(),
                     ))
-                    .layer(axum::extract::Extension(local_dataset_resolver)),
+                    .layer(axum::extract::Extension(catalog)),
```

Comment:
Note that I'm only passing `Catalog` now as an extension. 

This is consistent with GQL adapter. 

It allows to only construct repo instances when they are needed and gives objects back the control over their own lifetimes, so in future we could have flexibility of scoping instances per task / thread / singleton.